File: src/infinicore/ops/nll_loss/nll_loss.cc
#include "infinicore/ops/nll_loss.hpp"
#include "../../utils.hpp"

namespace infinicore::op {

common::OpDispatcher<NLLLoss::schema> &NLLLoss::dispatcher() {
    static common::OpDispatcher<NLLLoss::schema> dispatcher_;
    return dispatcher_;
};

void NLLLoss::execute(Tensor input, Tensor target, std::optional<Tensor> weight, Tensor output, int64_t ignore_index) {
    infinicore::context::setDevice(input->device());
    dispatcher().lookup(input->device().getType())(input, target, weight, output, ignore_index);
}

Tensor nll_loss(Tensor input, Tensor target, std::optional<Tensor> weight, int64_t ignore_index) {
    auto output = Tensor::empty({}, input->dtype(), input->device());
    nll_loss_(input, target, weight, output, ignore_index);
    return output;
}

void nll_loss_(Tensor input, Tensor target, std::optional<Tensor> weight, Tensor output, int64_t ignore_index) {
    NLLLoss::execute(input, target, weight, output, ignore_index);
}

} // namespace infinicore::op

File: python/infinicore/nn/functional/nll_loss.py
import infinicore
from infinicore.lib import _infinicore
from infinicore.tensor import Tensor


def nll_loss(
    input: Tensor,
    target: Tensor,
    weight: Tensor | None = None,
    ignore_index: int = -100,
    reduction: str = "mean",
    *,
    out=None,
) -> Tensor:
    if infinicore.use_ntops and input.device.type in ("cuda", "musa"):
        return infinicore.ntops.torch.nll_loss(
            input, target, weight=weight, ignore_index=ignore_index, reduction=reduction
        )

    weight_underlying = weight._underlying if weight is not None else None

    if out is None:
        return Tensor(
            _infinicore.nll_loss(
                input._underlying, target._underlying, weight_underlying, ignore_index
            )
        )

    _infinicore.nll_loss_(
        input._underlying,
        target._underlying,
        weight_underlying,
        out._underlying,
        ignore_index,
    )
    return out
